Add simulate erosion option to map generation

JanStrauss opened the pull request:
Add erosion simulation based on https://nickmcd.me/2020/04/10/simple-particle-based-hydraulic-erosion.
"Particles" are spawned at a random tile and will erode/deposit sediment from the heightmap based on velocity and slope. The position/velocity of the particles is updated using normals estimated from the heightmap by comparing the tile height with the neighbors.
The code is mostly based on the impl above but the normal calculation was changed, there is a directional bias in the original code from what I can tell. 
Most of the settings are not exposed to the ui, only the particle count as a particle per tile ratio to roughly have the same impact across different map sizes.
When the erosion option is enabled the weak smoothing is used which leads to more exposed edges but the strong version undoes most of the effect...

The map gen terrain page has a new Checkbox and spinner to enable the erosion feature and select how many particles to drop per tile:
<img width="402" alt="Screenshot 2025-08-19 at 15 35 46" src="https://github.com/user-attachments/assets/a9b25c34-c018-45c1-80c0-edddd2f4717f" />

Since this process can take some time (~seconds) a progress bar is shown while the erosion simulation runs:
<img width="612" alt="Screenshot 2025-08-19 at 15 38 19" src="https://github.com/user-attachments/assets/6e4b04fb-66b5-4e05-9d00-382d080129d8" />
It might be possible to apply the heightmap to the actual map and force a redraw every 10k particles to get a progress preview but that is maybe something for later.
The duration can actually vary depending on the terrain, e.g. with the island bias from https://github.com/OpenRCT2/OpenRCT2/pull/24946 the particles will run off the map and terminate the inner loop early while the valley bias will take longer as most particles will have to evaporate.

Example erosion result:
<img width="2048" height="1084" alt="Unnamed park 2025-08-02 08-31-39" src="https://github.com/user-attachments/assets/0f789583-18f1-418d-a7b2-de7817313c1d" />
<img width="2048" height="1084" alt="Unnamed park 2025-08-02 08-31-03" src="https://github.com/user-attachments/assets/769ec7e8-050e-4cb9-b47c-abda90f4cfb7" />

It might be nice to add an option to reduce the impact of particles one they are below the water level and expose the other config options.
There are some other possible follow ups e.g. wind erosion or integrated river generation:
https://nickmcd.me/2020/11/23/particle-based-wind-erosion/
https://nickmcd.me/2023/12/12/meandering-rivers-in-particle-based-hydraulic-erosion-simulations/

JanStrauss committed:
Add abstract Noise class, split SimplexFbmNoise from SimplexNoise

JanStrauss committed:
Change HeightMap height to float

JanStrauss committed:
Add VecXY and VecXYZ float vector types and offset method to TileCoordsXY

JanStrauss committed:
Add erosion simulation based on https://nickmcd.me/2020/04/10/simple-particle-based-hydraulic-erosion

The normal calculation is modified to fix a bias in contribution from the ordinal neighbors and simplified a bit. The cardinal/ordinal weights and density value were adjusted to compensate for the now on average larger non-z values.

JanStrauss committed:
Refactor simplex noise map gen and add erosion

- use floats when interacting with the heightmap
- instead of dividing the input heights and multiplying the heightmap value by 2 ensure the multiple-of-two requirement in a different way
- remove the heightmap oversampling and the more or less redundant slope calculation that was based on that
- support applying erosion
- Apply the heightmap random iteration gaussian smoothing thing only if erosion isn't selected
- Apply the weak non-heightmap smoothing function if erosion was used, keep the strong one otherwise
- move resetSurfaces just before the setMapHeight call to avoid artifacts with the progress window

JanStrauss committed:
Add erosion flag and particle per tile options to the gui

JanStrauss committed:
Update vcxproj

JanStrauss committed:
Show progress bar window when simulating erosion

JanStrauss committed:
Add erosion to png generator

ZehMatt commented on src/openrct2/world/map_generator/SimplexNoise.h:
```diff
@@ -9,14 +9,61 @@
+
 #include <cstdint>
 
+#ifdef __WARN_SUGGEST_FINAL_METHODS__
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wsuggest-final-methods"
+    #pragma GCC diagnostic ignored "-Wsuggest-final-types"
+#endif
```
Why do you need to ignore the warning if you could just add the final keyword to the classes?

JanStrauss replied:
IIRC only mingw complained about it and there are other locations that disable that warning

ZehMatt replied:
Well you did not mark the final class as final such as SimplexFbmNoise

ZehMatt commented on src/openrct2/world/Vec.hpp:
```diff
@@ -0,0 +1,192 @@
+
+#include <cmath>
+
+struct VecXY
```
Why not use CoordsXY? It is basically a vector just different name.

ZehMatt replied:
Hm right, I haven't really thought about that, I however honestly don't like that we have serval types doing the same thing essentially the only difference being the underlying type here. I wanted to have a base vector class for ages but there was push back and it got overruled to stick with the current code duplication, this however adds one more so its a bit of a weird situation.

ZehMatt commented on src/openrct2/world/Vec.hpp:
```diff
@@ -0,0 +1,192 @@
+    }
+};
+
+struct VecXYZ : VecXY
```
There is also CoordsXYZ

ZehMatt commented on src/openrct2/world/map_generator/Noise.h:
```diff
@@ -0,0 +1,20 @@
+
+namespace OpenRCT2::World::MapGenerator
+{
+    class Noise
```
Do we really need such interface when such abstraction is not even used in the code?

ZehMatt commented on src/openrct2/world/map_generator/Erosion.h:
```diff
@@ -0,0 +1,40 @@
+    struct ErosionSettings
+    {
+        ErosionSettings() = default;
+        ErosionSettings(Settings* settings)
```
```suggestion
        ErosionSettings(const Settings& settings)
```

ZehMatt commented on src/openrct2/world/map_generator/SimplexNoise.h:
```diff
@@ -9,14 +9,61 @@
+        float Generate(float x, float y) override;
+    };
+
+    class SimplexFbmNoise : public SimplexNoise
```
Another unnecessary abstraction, I don't see anything instantiate SimplexNoise.

JanStrauss replied:
See other comment, the wgen plugin branch exposes the non fbm noise through the api.
There are cases where you'd want to roll your own frequency summation instead of using fbm.

ZehMatt commented on src/openrct2/world/map_generator/Erosion.cpp:
```diff
@@ -0,0 +1,127 @@
+            position = _position;
+        }
+        VecXY position;
+        VecXY speed = VecXY(0.0f, 0.0f);
```
This should be called velocity, speed is typically a scalar unit.

ZehMatt commented on src/openrct2/world/map_generator/PngTerrainGenerator.cpp:
```diff
@@ -199,6 +205,7 @@ namespace OpenRCT2::World::MapGenerator
         const uint8_t rangeIn = maxValue - minValue;
         const uint8_t rangeOut = (settings->heightmapHigh - settings->heightmapLow) * 2;
 
+        MapInit({ mapHeight, mapWidth });
```
Why has this line been moved?

JanStrauss replied:
For the progress bar. This way the refreshed screen area around the progress bar window will still show the previous terrain instead of the newly initialized map

ZehMatt commented on src/openrct2/world/Location.hpp:
```diff
@@ -419,6 +419,11 @@ struct TileCoordsXY
         return rotatedCoords;
     }
 
+    constexpr TileCoordsXY Offset(int32_t _x, int32_t _y) const
```
Why is this needed? It already has operator+ implemented that does exactly this.

JanStrauss replied:
it isn't _needed_, slightly more readable imo. I can remove it.

ZehMatt commented on src/openrct2/world/map_generator/SimplexNoise.cpp:
```diff
@@ -25,34 +27,41 @@ namespace OpenRCT2::World::MapGenerator
+        std::mt19937 prng(seed);
+        for (auto& i : _perm)
+        {
+            i = prng() & 0xFF;
```
This is not ideal, use std::uniform_int_distribution.

JanStrauss committed:
address review feedback 1/n:

- rename speed to velocity
- change erosion settings constructor arg
- use uniform distribution to init simplex noise perm vector
- mark SimplexFbmNoise final
- remove offset from TileCoordsXY

misc:
- don't adjust surface level when below water level with erosion
- increase progress bar update rate and change message
- change default particles per tile to 3.00

JanStrauss commented:
I think I've addressed all your points apart from vec/coords and the noise classes. 
I'd also opt for a generic vec class but if that has been rejected previously I'd keep the two versions for now, should be fairly straightforward to refactor if the opinion on that changes later on. 
Re noise class hierarchy - I can remove it from this pr but as mentioned in the comments the plugin pr as well as the two other map gen prs sitting on top of this one make use of it so I'd prefer to keep it.

ZehMatt commented on src/openrct2/util/Util.cpp:
```diff
@@ -27,6 +27,14 @@ float UtilRandNormalDistributed()
     return _distributor(_prng);
 }
 
+uint8_t UtilRandUniformDistributedByte()
+{
+    thread_local std::mt19937 _prng{ std::random_device{}() };
+    // char types are not permitted as type params of uniform_int_distribution...
+    thread_local std::uniform_int_distribution<uint16_t> _distributor{ 0, 255 };
+    return static_cast<uint8_t>(_distributor(_prng));
+}
+
```
Move this into the Simplex class which is also the only class this will ever use this and it doesn't need to be thread_local lwhen its part of the class.

JanStrauss committed:
address review feedback 2/n:

- remove UtilRandUniformDistributedByte
- chain simplex noise constructors

misc:
- add const/explicit to erosion args/constructors

JanStrauss commented on src/openrct2/world/map_generator/SimplexNoise.cpp:
```diff
@@ -25,34 +27,43 @@ namespace OpenRCT2::World::MapGenerator
+    {
+        for (auto& i : _perm)
+        {
+            i = UtilRandUniformDistributedByte();
```
The constructor below allows to pass in a seed.  https://github.com/OpenRCT2/OpenRCT2/pull/24946 adds a map seed to the settings which is passed to all the prng/noise sources. The non seeded constructors are actually removed there.

ZehMatt reviewed (approved):
Code looks alright, I would like to have @AaronVanGeffen also take a look first.

AaronVanGeffen commented:
@JanStrauss Thanks for the PR. I'm very enthusiastic about your work on the map generator. Unfortunately, I have little time to dive in for a review this week. I hope to get to it by the weekend.

JanStrauss committed:
backport smoothHeightMap change from d9813f7cb3dac03f3d0e9e64efc04ce16ef6f131.

Fixes artifacts on the map border that previously were (mostly) masked by the 2x oversampling.

JanStrauss commented:
Sorry for the late response, didn't see your edit.

> 'Simulate erosion' is a bit broad in meaning, right? We can probably do better, but I'm not sure how yet

Yep it's pretty broad, but I couldn't think of something more fitting. Land erosion seemed a bit redundant and calling it hydraulic erosion felt over specific without wind or thermal erosion being implemented... 

> I realise you have an extensive follow-up PR in the pipeline as well, with more landscape biases, though, right? How does that tie in to this one?

They are independent features, you can use the biases without erosion and vice versa. The biases are only available with simplex/warped noise atm (could also be added to the png heightmap source but didn't seem like a likely use case) and modify the normalized noise value before it's scaled by terrain height. Erosion operates on the heightmap and I see it more as as a post-processing step similar to the smoothing options.
